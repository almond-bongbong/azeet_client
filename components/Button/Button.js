import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import SimpleLoader from 'components/Loader/SimpleLoader';
import styled, { css } from 'styled-components';
import { red, blue } from 'style/mixin';

const BasicButtonStyle = styled.div`
  display: block;
  
  ${props => props.inline && css`
    display: inline-block;
    width: auto;
  `};

  button, a {
    display: block;
    width: ${props => `${props.width !== 'auto' ? `${props.width}px` : '100%'}`};
    min-width: 80px;
    height: ${props => props.height}px;
    padding: 0 20px;
    border: 1px solid #ddd;
    border-radius: 2px;
    outline: 0;
    font-size: 13px;
    cursor: pointer;
    text-align: center;
    text-decoration: none;
    line-height: ${props => props.height}px;
    transition: box-shadow .2s;
  
    &:hover {
      box-shadow: 1px 1px 3px rgba(0,0,0,0.2);
    }
    
    &:active {
      box-shadow: 0 0 2px rgba(0,0,0,0.1);
    }
    
    & span {
      position: relative;
    }
    
    ${props => props.theme === 'white' && css`
      background-color: #fff;
      border: 1px solid #ddd;
      color: #666;
    `};
    
    ${props => props.theme === 'gray' && css`
      background-color: #666;
      border: 1px solid #666;
      color: #fff;
    `}
    
    ${props => props.theme === 'red' && css`
      background-color: ${red};
      border: 1px solid ${red};
      color: #fff;
    `};
    
    ${props => props.theme === 'blue' && css`
      background-color: ${blue};
      border: 1px solid ${blue};
      color: #fff;
    `}
  }
`;

const Button = ({
  text, theme, width, height, inline, loading, onClick,
}) => {
  const loaderTheme = useMemo(() => (theme === 'white' ? 'red' : 'white'), [theme]);
  const loaderSize = useMemo(() => height / 2, [height]);

  return (
    <BasicButtonStyle data-testid="button-wrap" theme={theme} width={width} height={height} inline={inline}>
      {loading ? (
        <button type="button" title="로딩중"><SimpleLoader theme={loaderTheme} size={loaderSize} /></button>
      ) : (
        <button type="button" onClick={onClick}><span>{text}</span></button>
      )}
    </BasicButtonStyle>
  );
};

Button.propTypes = {
  text: PropTypes.string,
  inline: PropTypes.bool,
  theme: PropTypes.oneOf(['white', 'gray', 'red', 'blue']),
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  height: PropTypes.oneOf([30, 40, 50]),
  loading: PropTypes.bool,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  text: '버튼',
  theme: 'white',
  width: 'auto',
  height: 40,
  inline: true,
  loading: false,
  onClick: undefined,
};

export default Button;
