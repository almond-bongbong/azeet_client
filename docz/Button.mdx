---
name: Button
---

import { Playground, Props } from 'docz'
import Button from 'components/Button/Button'

# Button

<Props of={Button} />

## basic usage

<Playground>
  <Button>Button</Button>
</Playground>

## inline

### true

<Playground>
  <Button theme="red">버튼</Button>
</Playground>

### false

<Playground>
  <Button theme="red" inline={false}>버튼</Button>
</Playground>

## theme

### white

<Playground>
  <Button theme="white">취소</Button>
</Playground>

### gray

<Playground>
  <Button theme="gray">취소</Button>
</Playground>

### red

<Playground>
  <Button theme="red">버튼</Button>
</Playground>

### blue

<Playground>
  <Button theme="blue">버튼</Button>
</Playground>

## width

<Playground>
  <Button width={120} theme="white">작은거</Button>
</Playground>

## height

### 30

<Playground>
  <Button height={30} theme="white">작은거</Button>
</Playground>

### 40

<Playground>
  <Button height={40} theme="white">보통</Button>
</Playground>

### 50

<Playground>
  <Button height={50} theme="white">큰거</Button>
</Playground>

## loading

<Playground>
  <Button loading={true}>로딩</Button>
</Playground>

## event

### onClick(event)

<Playground>
  <Button theme="white" onClick={e => alert('클릭')}>클릭!</Button>
</Playground>