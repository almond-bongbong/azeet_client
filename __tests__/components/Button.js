import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Button from 'components/Button/Button';

describe('Button', () => {
  it('matches snapshot', () => {
    const utils = render(<Button />);

    expect(utils.container).toMatchSnapshot();
  });

  it('calls onClick', () => {
    const onClick = jest.fn();
    const { getByText } = render(<Button onClick={onClick} text="버튼" />);
    const button = getByText('버튼');

    fireEvent.click(button);

    expect(onClick).toBeCalled();
  });

  it('not calls onClick with while loading', () => {
    const onClick = jest.fn();
    const { getByTitle } = render(<Button onClick={onClick} loading text="로딩버튼" />);
    const button = getByTitle('로딩중');

    fireEvent.click(button);

    expect(onClick).not.toBeCalled();
  });
});
