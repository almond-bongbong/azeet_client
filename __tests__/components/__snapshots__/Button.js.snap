// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<div>
  <div
    class="Button__BasicButtonStyle-f73wfv-0 jSeGVY"
    data-testid="button-wrap"
    height="40"
    width="auto"
  >
    <button
      type="button"
    >
      <span>
        버튼
      </span>
    </button>
  </div>
</div>
`;
